File: chrome-extension/lib/transcript-turns.js
/**
 * Parse / normalize diarized transcripts for read-only review.
 */

const SPEAKER_LINE = /^(?:SPEAKER:\s*)?(S\d+|Speaker\s*\d+)\s*:?\s*$/i;
const SPEAKER_INLINE = /^(?:SPEAKER:\s*)?(S\d+|Speaker\s*\d+)\s*[:.-]\s*(.+)$/i;
const NAMED_SPEAKER_LINE = /^SPEAKER:\s*(?!S\d+\b)(.+?)\s*$/i;
const NAMED_SPEAKER_INLINE = /^SPEAKER:\s*(?!S\d+\b)([^:]+?)\s*[:.-]\s*(.+)$/i;

export function normalizeSpeaker(raw) {
  const m = String(raw || '').match(/(\d+)/);
  return m ? `S${m[1]}` : (raw ? String(raw).trim().toUpperCase() : null);
}

export function speakerSide(raw) {
  const s = normalizeSpeaker(raw);
  if (s === 'S1') return 's1';
  if (s === 'S2') return 's2';
  // Named diarization (JUAN) on a 2-party call sits with the prospect.
  if (s && !/^S\d+$/.test(s)) return 's2';
  return 'other';
}

export function speakerDisplayLabel(raw, labels = {}) {
  const side = speakerSide(raw);
  if (side === 's1') return labels.s1 || 'You';
  if (side === 's2') return labels.s2 || 'Them';
  return labels.s2 || raw || 'Speaker';
}

export function firstName(full) {
  const part = String(full || '').trim().split(/\s+/)[0];
  return part || '';
}

export function parseTranscriptTurns(text) {
  const raw = String(text || '').trim();
  if (!raw) return [];

  const turns = [];
  let current = null;

  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed) {
      if (current) current.text += '\n';
      continue;
    }
    const inline = trimmed.match(SPEAKER_INLINE);
    if (inline) {
      if (current && current.text.trim()) turns.push(current);
      current = { speaker: inline[1].replace(/\s+/g, ''), text: inline[2].trim() };
      continue;
    }
    const onlySpeaker = trimmed.match(SPEAKER_LINE);
    if (onlySpeaker) {
      if (current && current.text.trim()) turns.push(current);
      current = { speaker: onlySpeaker[1].replace(/\s+/g, ''), text: '' };
      continue;
    }
    const namedInline = trimmed.match(NAMED_SPEAKER_INLINE);
    if (namedInline) {
      if (current && current.text.trim()) turns.push(current);
      current = { speaker: namedInline[1].trim(), text: namedInline[2].trim() };
      continue;
    }
    const namedOnly = trimmed.match(NAMED_SPEAKER_LINE);
    if (namedOnly) {
      if (current && current.text.trim()) turns.push(current);
      current = { speaker: namedOnly[1].trim(), text: '' };
      continue;
    }
    if (!current) current = { speaker: null, text: trimmed };
    else current.text = current.text ? `${current.text}\n${trimmed}` : trimmed;
  }
  if (current && current.text.trim()) turns.push(current);
  if (!turns.length && raw) return [{ speaker: null, text: raw }];
  return turns
    .map((t) => ({ speaker: t.speaker, text: t.text.replace(/\n+$/g, '').trim() }))
    .filter((t) => t.text);
}

function fingerprint(turn) {
  const speaker = normalizeSpeaker(turn.speaker) || '';
  const text = String(turn.text || '').replace(/\s+/g, ' ').trim().toLowerCase();
  return `${speaker}:${text.slice(0, 160)}`;
}

export function dedupeRepeatedConversation(turns) {
  if (!turns || turns.length < 4) return turns || [];
  const fps = turns.map(fingerprint);
  const n = turns.length;
  if (n % 2 === 0) {
    const mid = n / 2;
    if (fps.slice(0, mid).join('|') === fps.slice(mid).join('|')) {
      return turns.slice(0, mid);
    }
  }
  return turns;
}

export function mergeConsecutiveTurns(turns) {
  const out = [];
  for (const turn of turns || []) {
    const speaker = normalizeSpeaker(turn.speaker);
    const text = String(turn.text || '').trim();
    if (!text) continue;
    if (out.length && normalizeSpeaker(out[out.length - 1].speaker) === speaker) {
      out[out.length - 1].text = `${out[out.length - 1].text}\n${text}`;
    } else {
      out.push({ speaker, text });
    }
  }
  return out;
}

export function serializeTranscriptTurns(turns) {
  return (turns || [])
    .map((turn) => {
      const text = String(turn.text || '').trim();
      if (!text) return '';
      const speaker = normalizeSpeaker(turn.speaker);
      return speaker ? `SPEAKER: ${speaker}\n${text}` : text;
    })
    .filter(Boolean)
    .join('\n\n');
}

export function normalizeDiarizedTranscript(text) {
  const turns = mergeConsecutiveTurns(dedupeRepeatedConversation(parseTranscriptTurns(text)));
  if (!turns.length) return String(text || '').trim();
  return serializeTranscriptTurns(turns);
}

File: backend/app/services/transcript_turns.py
"""Parse, dedupe, and serialize diarized call transcripts."""

from __future__ import annotations

import re
from typing import Optional

_SPEAKER_LINE = re.compile(
    r"^(?:SPEAKER:\s*)?(S\d+|Speaker\s*\d+)\s*:?\s*$",
    re.IGNORECASE,
)
_SPEAKER_INLINE = re.compile(
    r"^(?:SPEAKER:\s*)?(S\d+|Speaker\s*\d+)\s*[:.-]\s*(.+)$",
    re.IGNORECASE,
)


def normalize_speaker(raw: Optional[str]) -> Optional[str]:
    if not raw:
        return None
    m = re.search(r"(\d+)", str(raw))
    return f"S{m.group(1)}" if m else str(raw).strip().upper()


def parse_transcript_turns(transcript: str) -> list[dict]:
    """Parse Speechmatics-style SPEAKER: S1 / Speaker 1 blocks into turns."""
    raw = (transcript or "").strip()
    if not raw:
        return []

    turns: list[dict] = []
    current: Optional[dict] = None
    for line in raw.splitlines():
        trimmed = line.strip()
        if not trimmed:
            if current and current.get("text"):
                current["text"] += "\n"
            continue
        inline = _SPEAKER_INLINE.match(trimmed)
        if inline:
            if current and (current.get("text") or "").strip():
                turns.append(current)
            current = {"speaker": inline.group(1), "text": inline.group(2).strip()}
            continue
        only = _SPEAKER_LINE.match(trimmed)
        if only:
            if current and (current.get("text") or "").strip():
                turns.append(current)
            current = {"speaker": only.group(1), "text": ""}
            continue
        if not current:
            current = {"speaker": None, "text": trimmed}
        else:
            current["text"] = f"{current['text']}\n{trimmed}" if current["text"] else trimmed

    if current and (current.get("text") or "").strip():
        turns.append(current)

    if not turns and raw:
        return [{"speaker": None, "text": raw}]

    out = []
    for t in turns:
        text = re.sub(r"\n+$", "", (t.get("text") or "")).strip()
        if text:
            out.append({"speaker": t.get("speaker"), "text": text})
    return out


def _fingerprint(turn: dict) -> str:
    speaker = normalize_speaker(turn.get("speaker")) or ""
    text = re.sub(r"\s+", " ", (turn.get("text") or "")).strip().lower()
    return f"{speaker}:{text[:160]}"


def dedupe_repeated_conversation(turns: list[dict]) -> list[dict]:
    """Drop a second copy of the same call (Speaker 1 view + SPEAKER: S1 raw)."""
    if len(turns) < 4:
        return turns
    fps = [_fingerprint(t) for t in turns]
    n = len(turns)
    if n % 2 == 0:
        mid = n // 2
        if fps[:mid] == fps[mid:]:
            return turns[:mid]
    for size in range(2, n // 2 + 1):
        if fps[:size] == fps[size : size * 2] and size * 2 == n:
            return turns[:size]
    return turns


def merge_consecutive_turns(turns: list[dict]) -> list[dict]:
    out: list[dict] = []
    for turn in turns:
        speaker = normalize_speaker(turn.get("speaker"))
        text = (turn.get("text") or "").strip()
        if not text:
            continue
        if out and normalize_speaker(out[-1].get("speaker")) == speaker:
            out[-1]["text"] = f"{out[-1]['text']}\n{text}"
        else:
            out.append({"speaker": speaker, "text": text})
    return out


def serialize_transcript_turns(turns: list[dict]) -> str:
    parts: list[str] = []
    for turn in turns:
        text = (turn.get("text") or "").strip()
        if not text:
            continue
        speaker = normalize_speaker(turn.get("speaker"))
        if speaker:
            parts.append(f"SPEAKER: {speaker}\n{text}")
        else:
            parts.append(text)
    return "\n\n".join(parts)


def normalize_diarized_transcript(transcript: str) -> str:
    turns = merge_consecutive_turns(
        dedupe_repeated_conversation(parse_transcript_turns(transcript))
    )
    if not turns:
        return (transcript or "").strip()
    return serialize_transcript_turns(turns)
